File: C++/LinkedListCycle.cc
// Given a linked list, determine if it has a cycle in it.

// Follow up:
// Can you solve it without using extra space? 

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast && fase->next) {
        	if (fast == slow) {
        		return true;
        	}
        	fast = fast->next->next;
        	slow = slow->next;
        }
        return false;
    }
};

File: Java/LinkedListCycle.java
// Given a linked list, determine if it has a cycle in it.

public class LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
            if (fast == null) {
                break;
            }
            fast = fast.next;
            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
}
